### PR TITLE
From_graph impls

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by CodeRabbit

- **New Features**
  - Introduced new methods to create various graph representations directly from existing graphs, enabling seamless conversion between graph types such as adjacency lists, edge lists, edge-node lists, and matrix-based graphs.
  - Added a new example demonstrating graph type conversions and usage scenarios.

- **Bug Fixes**
  - Improved error handling for capacity limits and invalid node indices during graph conversions.

- **Tests**
  - Added extensive tests to verify correctness of graph conversions, error conditions, and edge cases.

- **Documentation**
  - Enhanced documentation for new conversion methods, including usage constraints and examples.

### CODE REVIEW

File: src/adjacency_list/slice_adjacency_list.rs [hunk 1]
```diff
@@ -128,6 +128,84 @@ where
...
+    /// // Create a source graph (adjacency list)
+    /// let mut dict = Dictionary::<usize, [usize; 2], 8>::new();
+    /// dict.insert(0, [1, 2]).unwrap();
+    /// dict.insert(1, [2, 0]).unwrap();
+    /// dict.insert(2, [0, 1]).unwrap();
+    /// let source = MapAdjacencyList::new_unchecked(dict);
+    ///
+    /// // Convert to SliceAdjacencyList with exactly 3 nodes and capacity for edges
+    /// let slice_graph: SliceAdjacencyList<usize, NodeStructOption<4, usize>, [(usize, NodeStructOption<4, usize>); 3]> =
+    ///     SliceAdjacencyList::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/adjacency_list/slice_adjacency_list.rs]: @gemini-code-assist doublecheck this - i think i fixed it

File: src/edgelist/edge_list.rs [hunk 2]
```diff
@@ -52,6 +52,62 @@ impl<const N: usize, NI, E> EdgeList<N, NI, E> {
...
+    ///
+    /// // Create a source graph (adjacency list)
+    /// let mut dict = Dictionary::<usize, [usize; 2], 8>::new();
+    /// dict.insert(0, [1, 2]).unwrap();
+    /// dict.insert(1, [2, 0]).unwrap();
+    /// let source = MapAdjacencyList::new_unchecked(dict);
+    ///
+    /// // Convert to EdgeList
+    /// let edge_list: EdgeList<8, usize, EdgeStructOption<16, usize>> =
+    ///     EdgeList::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/edgelist/edge_list.rs]: @gemini-code-assist doublecheck this - i think i fixed it

File: src/edgelist/edge_node_list.rs [hunk 3]
```diff
@@ -46,6 +46,77 @@ impl<NI, E, N> EdgeNodeList<NI, E, N> {
...
+    /// // Create a source graph (adjacency list)
+    /// let mut dict = Dictionary::<usize, [usize; 2], 8>::new();
+    /// dict.insert(0, [1, 2]).unwrap();
+    /// dict.insert(1, [2, 0]).unwrap();
+    /// dict.insert(2, [0, 1]).unwrap();
+    /// let source = MapAdjacencyList::new_unchecked(dict);
+    ///
+    /// // Convert to EdgeNodeList with capacity for nodes and edges
+    /// let edge_node_graph: EdgeNodeList<usize, EdgeStructOption<8, usize>, NodeStructOption<4, usize>> =
+    ///     EdgeNodeList::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/edgelist/edge_node_list.rs]: @gemini-code-assist doublecheck this - i think i fixed it

File: src/matrix/bit_matrix.rs [hunk 4]
```diff
@@ -88,6 +88,75 @@ impl<const C: usize, const R: usize> BitMatrix<C, R> {
...
+    ///
+    /// // Create a source graph (adjacency list)
+    /// let mut dict = Dictionary::<usize, [usize; 2], 8>::new();
+    /// dict.insert(0, [1, 2]).unwrap();
+    /// dict.insert(1, [2, 0]).unwrap();
+    /// dict.insert(2, [0, 1]).unwrap();
+    /// let source = MapAdjacencyList::new_unchecked(dict);
+    ///
+    /// // Convert to BitMatrix (1 column for 8 nodes: 8 = 8*1)
+    /// let matrix: BitMatrix<1, 8> = BitMatrix::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/bit_matrix.rs]: @gemini-code-assist doublecheck this - i think i fixed it

File: src/matrix/simple_matrix.rs [hunk 5]
```diff
@@ -20,7 +20,91 @@
...
+    /// // Create a source graph (adjacency list)
+    /// let mut dict = Dictionary::<usize, [usize; 2], 8>::new();
+    /// dict.insert(0, [1, 2]).unwrap();
+    /// dict.insert(1, [2, 0]).unwrap();
+    /// dict.insert(2, [0, 1]).unwrap();
+    /// let source = MapAdjacencyList::new_unchecked(dict);
+    ///
+    /// // Convert to Matrix (3x3 matrix to fit nodes 0, 1, 2)
+    /// let matrix: Matrix<3, (), [[Option<()>; 3]; 3], [Option<()>; 3]> =
+    ///     Matrix::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/simple_matrix.rs]: @gemini-code-assist doublecheck this - i think i fixed it

File: src/matrix/map_matrix.rs [hunk 6]
```diff
@@ -53,6 +53,132 @@ where
...
+    /// use heapless_graphs::edges::EdgeStructOption;
+    /// use heapless_graphs::containers::maps::staticdict::Dictionary;
+    ///
+    /// // Create a source graph (edge list)
+    /// let edges = EdgeStructOption([Some((0, 1)), Some((1, 2)), Some((0, 2)), None]);
+    /// let source = EdgeList::<4, usize, _>::new(edges);
+    ///
+    /// // Convert to MapMatrix (3x3 matrix to fit nodes 0, 1, 2)
+    /// let map_matrix: MapMatrix<3, usize, (), Dictionary<usize, usize, 8>, [[Option<()>; 3]; 3], [Option<()>; 3]> =
+    ///     MapMatrix::from_graph(&source).unwrap();
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/map_matrix.rs]: It can't infer that much.

File: src/matrix/bit_map_matrix.rs [hunk 7]
```diff
@@ -55,6 +55,121 @@ where
...
+    /// * `source_graph` - The graph to copy nodes and edges from
+    ///
+    /// # Returns
+    /// * `Ok(BitMapMatrix)` if successful
+    /// * `Err(GraphError)` if too many nodes or iteration fails
+    ///
+    /// # Constraints
+    /// * Source graph must have at most 8*N nodes (BitMatrix capacity)
+    /// * Node index type must implement Copy
+    /// * Requires R = 8*N for valid BitMatrix dimensions
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/bit_map_matrix.rs:79]: I added a test for this - it's actually checked by ::new transitively

File: src/matrix/bit_map_matrix.rs [hunk 8]
```diff
@@ -55,6 +55,96 @@ where
...
+                    index_map
+                        .insert(node, matrix_index)
+                        .map_err(|_| GraphError::OutOfCapacity)?;
+                }
+            }
+            Err(_) => {
+                // If iter_nodes() fails (e.g., empty EdgeList), we'll create an empty matrix
+                // This is valid - the matrix will have no nodes mapped
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/bit_map_matrix.rs]: Double check it please
